pdp/handlers_create.go

```diff
@@ -4,16 +4,150 @@ import (
 	"encoding/hex"
 	"encoding/json"
 	"io"
+	"math/big"
 	"net/http"
 	"path"
 	"strings"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 
+	"github.com/filecoin-project/curio/harmony/harmonydb"
 	"github.com/filecoin-project/curio/pdp/contract"
 )
 
+// handleCreateDataSetAndAddPieces handles the creation of a new data set and adding pieces at the same time
+func (p *PDPService) handleCreateDataSetAndAddPieces(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// Step 1: Verify that the request is authorized using ECDSA JWT
+	serviceLabel, err := p.AuthService(r)
+	if err != nil {
+		http.Error(w, "Unauthorized: "+err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	type RequestBody struct {
+		RecordKeeper string            `json:"recordKeeper"`
+		Pieces       []AddPieceRequest `json:"pieces"`
+		ExtraData    *string           `json:"extraData,omitempty"`
+	}
+
+	var reqBody RequestBody
+	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
+		http.Error(w, "Invalid JSON in request body: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if reqBody.RecordKeeper == "" {
+		http.Error(w, "recordKeeper address is required", http.StatusBadRequest)
+		return
+	}
+
+	recordKeeperAddr := common.HexToAddress(reqBody.RecordKeeper)
+	if recordKeeperAddr == (common.Address{}) {
+		http.Error(w, "Invalid recordKeeper address", http.StatusBadRequest)
+		return
+	}
+
+	// Check if the recordkeeper is in the whitelist for public services
+	if contract.IsPublicService(serviceLabel) && !contract.IsRecordKeeperAllowed(recordKeeperAddr) {
+		http.Error(w, "recordKeeper address not allowed for public service", http.StatusForbidden)
+		return
+	}
+
+	extraDataBytes, err := decodeExtraData(reqBody.ExtraData)
+	if err != nil {
+		http.Error(w, "Invalid extraData format (must be hex encoded)", http.StatusBadRequest)
+		return
+	}
+
+	pieceDataArray, subPieceInfoMap, _, err := p.transformAddPiecesRequest(ctx, serviceLabel, reqBody.Pieces)
+	if err != nil {
+		http.Error(w, "Failed to transform addPieces request: "+err.Error(), http.StatusBadRequest)
```

Comment:
```suggestion
		http.Error(w, "Failed to process request: "+err.Error(), http.StatusBadRequest)
```

"transform" isn't helpful to the user here I think